### PR TITLE
distribution: errors: do not retry if too many login attempts

### PR DESCRIPTION
**Pull request by runcom**
Hit this while looking for other stuff in the push code. Basically this stops retrying if that too many login attempts error is encountered (I don't see why keep trying to login, you'll wait more and more.). I believe this happens only against the official docker hub though. The check isn't perfect, if you have suggestions, welcome. I thought this string check could be avoided by placing this error in `errcode` pkg in distribution, again, not sure.

ping @aaronlehmann 

(to @aaronlehmann as a side note, the client doesn't receive the error, as the push with unauthorized case from https://github.com/docker/docker/pull/22253#issuecomment-213470857, the client just gets 200 because I think the status code is written before to allow streaming of chars, not sure)

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
@runcom should this fix https://github.com/docker/docker/issues/22188#issuecomment-213635233 ?

File: distribution/errors.go [hunk 1]
```diff
@@ -92,6 +92,9 @@ func retryOnError(err error) error {
 		case errcode.ErrorCodeUnauthorized, errcode.ErrorCodeUnsupported, errcode.ErrorCodeDenied:
 			return xfer.DoNotRetry{Err: err}
 		}
+		if strings.Contains(v.Error(), "too many failed login attempts for username or IP address") {
```
Comment (thaJeztah) [inline distribution/errors.go]: Instinctively, I'd be a bit less specific, perhaps just check for "too many failed login attempts", but not sure if these messages ever change

Comment (runcom) [inline distribution/errors.go]: Not sure, this error probably comes from just the official docker hub

**Comment by runcom** [comment]
I don't think this fixes that error though. This just stops Retrying if this error is encountered.

File: distribution/errors.go (hunk 1 above)
Comment (thaJeztah) [inline distribution/errors.go]: Yeah, thought that just that bit may be specific enough, but it's a nit

**Comment by thaJeztah** [comment]
Yeah, you're right. Don't think there actually *is* a solution for that, if different pull/push sessions attempt to do so with wrong credentials, then a username / IP-address will still be blocked

File: distribution/errors.go (hunk 1 above)
Comment (Somebi) [inline distribution/errors.go]: It's coming from official docker hub.

**Comment by Somebi** [comment]
You just missing some good test for push operation. Including throttling test, connection lost tests etc.

**Comment by Somebi** [comment]
Some networth or auth limit tests.

**Comment by runcom** [comment]
Not sure we have that many tests again the official docker hub. And not sure how to test this behavior honestly.

**Comment by Somebi** [comment]
Maybe initial login attempt freezed for a while due to some inner docker hub network issues and for that time client send a lot of other login attempts. And thus reaching login attempts quota. At what stage doea login attempt is incremented and is counted as successful retry?

**Comment by Somebi** [comment]
Create various servers in various datacenters. Hardcode them in tests. Few clients and a hub server. Make login attempts to a hub in a various scenarios and register expected response codes, timings, timeouts as successful.

**Comment by aaronlehmann** [comment]
@runcom: Do you know what HTTP error code the hub returns in this case? If it's something specific like `429 Too many requests`, I'd rather check for the code than the string.

**Comment by runcom** [comment]
@aaronlehmann that makes sense. I haven't checked but I will and update the code to check the error.

**Comment by runcom** [comment]
@aaronlehmann it's not a 429 :confused:

**Comment by aaronlehmann** [comment]
LGTM

**Comment by cpuguy83** [comment]
Honestly not a big fan of this change. I don't like that we are inspecting hub specific error messages (which aren't guaranteed to be stable).

**Comment by Somebi** [comment]
Well at least respond to client properly. Otherwise it just spam auth
servers...

**Comment by runcom** [comment]
moved back to design review, not sure what to do then..I agree with @Somebi though

**Comment by aaronlehmann** [comment]
I will reach out to the hub team and ask if it's possible for them to use a `429` error code for the "too many retries" error.

**Comment by aaronlehmann** [comment]
@runcom: I checked with Hub and they think it is using the 429 error code for this condition. What code were you seeing instead? If it's not 429, that sounds like a bug that we can get fixed.

**Comment by runcom** [comment]
@aaronlehmann I've opened https://github.com/docker/distribution/pull/1693 to check for 429, in order for this PR to be merged we should re-vendor docker/distribution code.

**Comment by aaronlehmann** [comment]
@runcom: Sounds good. Is the hub returning 429 after all?

**Comment by runcom** [comment]
@aaronlehmann yup

**Comment by runcom** [comment]
@aaronlehmann done - split the commits cause the vendoring doesn't affect the code at all - then second commit actually fixes the issue in question.

**Comment by cpuguy83** [comment]
Much better!

LGTM

**Comment by aaronlehmann** [comment]
LGTM

**Comment by thaJeztah** [comment]
nice, much better indeed!

**Comment by thaJeztah** [comment]
@runcom looks like CI is failing;

```
17:59:14 ---> Making bundle: binary-client (in bundles/1.12.0-dev/binary-client)
17:59:14 Building: bundles/1.12.0-dev/binary-client/docker-1.12.0-dev
17:59:25 # github.com/docker/distribution/registry/api/errcode
17:59:25 vendor/src/github.com/docker/distribution/registry/api/errcode/register.go:74: undefined: http.StatusTooManyRequests
```

**Comment by runcom** [comment]
@thaJeztah WAT?!

**Comment by cpuguy83** [comment]
Originall was thinking maybe it was defined in go 1.5, but sure enough it's there.

**Comment by runcom** [comment]
@cpuguy83 it's not into go1.5.x, it's only available into go1.6 - so tests in docker/distribution passed but here are failing...

**Comment by cpuguy83** [comment]
Apparently I'm running godoc against 1.6 somehow... source code shows it as unexported for 1.5.. so yeah, makes sense.

**Comment by runcom** [comment]
I'll revendor after https://github.com/docker/distribution/pull/1696 is merged (hopefully)

**Comment by runcom** [comment]
I've revendored docker/distribution, moving to merge again, hopefully tests will be ok

**Comment by vdemeester** [comment]
@runcom yay everything is back to :green_heart:, merging :wink:

**Comment by runcom** [comment]
:tada:

**Pull request merged**